### PR TITLE
 docs: Transaction consensus rules: Header rules

### LINKED ISSUE #3425: Document consensus rules from Zcash spec 7.1.2: Transaction Header
- [x] The transaction version number MUST be greater than or equal to 1.
- [x] [Pre-Overwinter] The `fOverwintered` fag MUST NOT be set.
- [x] [Overwinter onward] The `fOverwintered` flag MUST be set.
- [x] [Overwinter onward] The version group ID MUST be recognized.
- [x] [Overwinter only, pre-Sapling] The transaction version number MUST be 3, and the version group ID MUST be 0x03C48270.
- [x] [Sapling to Canopy inclusive, pre-NU5] The transaction version number MUST be 4, and the version group ID MUST be 0x892F2085.
- [x] [NU5 onward] The transaction version number MUST be 4 or 5. If the transaction version number is 4 then the version group ID MUST be 0x892F2085. If the transaction version number is 5 then the version group ID MUST be 0x26A7270A.
- [x] [NU5 onward] The `nConsensusBranchId` field MUST match the consensus branch ID used for SIGHASH transaction hashes, as specified in [ZIP-244].
- [x] [Overwinter to Canopy inclusive, pre-NU5] `nExpiryHeight` MUST be less than or equal to 499999999.
- [x] [NU5 onward] `nExpiryHeight` MUST be less than or equal to 499999999 for non-coinbase transactions.
- [x] [Overwinter onward] If a transaction is not a coinbase transaction and its `nExpiryHeight` field is nonzero, then it MUST NOT be mined at a block height greater than its `nExpiryHeight`.

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

The number of consensus rules related to transactions is huge so we splitted into 4 categories. More details at https://github.com/ZcashFoundation/zebra/issues/3223

I made the split and identified 11 rules for the header section. If merged this will close https://github.com/ZcashFoundation/zebra/issues/3425

## Solution

Consensus rules were documented according to the format the zebra team is implementing for all consensus rules. 

Some duplicated docs were removed so the exact consensus rule docs can be found only once in the codebase (where possible), please make sure no doc comment that you think clarifies was removed. 

## Review

Anyone can review.

### Reviewer Checklist

  - [x] Consensus rule docs are in a good spot.
  - [x] All the rules in the issue were added of modified in this PR.
  - [ ] No piece of doc comment that i want to keep was deleted.

## Follow Up Work

The other 3 sections of the transaction consensus rules which are in the following issues:

- https://github.com/ZcashFoundation/zebra/issues/3426
- https://github.com/ZcashFoundation/zebra/issues/3427
- https://github.com/ZcashFoundation/zebra/issues/3428

### CODE REVIEW

File: zebra-chain/src/transaction/serialize.rs [hunk 1]
```diff
@@ -544,10 +544,33 @@ impl ZcashDeserialize for Transaction {
...
+        //
+        // > [Overwinter onward] The fOverwintered flag MUST be set.
+        //
+        // > [Overwinter only, pre-Sapling] The transaction version number MUST be 3,
+        // > and the version group ID MUST be 0x03C48270.
+        //
+        // > [Sapling to Canopy inclusive, pre-NU5] The transaction version number MUST be 4,
+        // > and the version group ID MUST be 0x892F2085.
+        //
+        // > [NU5 onward] The transaction version number MUST be 4 or 5.
```
Comment (conradoplg) [inline zebra-chain/src/transaction/serialize.rs:571]: I think this is not checked here (it parses whatever version the transaction is), neither where the old comment was. We kind of test it in a roundabout way.

Here we only allow version 4 or 5 (other versions are checkpointed and never fully verified). This already satisfies the consensus rule.

https://github.com/ZcashFoundation/zebra/blob/154fc0a2fb01a5d9d45f0cb5d3ebc25b5624dc4b/zebra-consensus/src/transaction.rs#L360

The place where the old comment was checks that, for example, if transaction is 4, then it must be Sapling-onward. Which is not exactly the same thing.

What do you think of moving this rule to the line above?

**Review by conradoplg — CHANGES_REQUESTED** [review]
Looks good, I was as thorough as possible but only found one rule that seems to fit better in another place.

File: zebra-chain/src/transaction/serialize.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-chain/src/transaction/serialize.rs:571]: Hmm, I am not so sure about it. It is true that here we are not checking if the transaction is 4 or 5 but when we get a transaction 4 or 5 we make sure the group id is correct which is the second part of those 2 rules.

Comment (oxarbitrage) [reply zebra-chain/src/transaction/serialize.rs:571]: What do you think about something like https://github.com/ZcashFoundation/zebra/pull/3456/commits/124760e4ffe821caee60830e79cfbd8231690a30 ?

Comment (conradoplg) [reply zebra-chain/src/transaction/serialize.rs:571]: Looks good!

**Review by conradoplg — DISMISSED** [review]

File: zebra-chain/src/transaction/serialize.rs (hunk 1 above)
Comment (teor2345) [reply zebra-chain/src/transaction/serialize.rs:571]: Checkpointing allows us to ignore most rules for pre-Canopy network upgrades.

But we still need to check that Canopy and NU5 have the correct transaction versions. That is done here:

https://github.com/ZcashFoundation/zebra/blob/fa071562fd0b891f29143e4cc8d17b34b4b1cf10/zebra-consensus/src/transaction.rs#L547-L550

https://github.com/ZcashFoundation/zebra/blob/fa071562fd0b891f29143e4cc8d17b34b4b1cf10/zebra-consensus/src/transaction.rs#L636-L637

Comment (teor2345) [reply zebra-chain/src/transaction/serialize.rs:571]: In particular, Canopy can't have version 5 transactions, so we still need to check these rules somewhere. (Parsing and requiring 4 or 5 for sighash are not enough.)

Comment (oxarbitrage) [reply zebra-chain/src/transaction/serialize.rs:571]: So i think we should keep those docs. I re added them updating the format at https://github.com/ZcashFoundation/zebra/pull/3456/commits/eb20e78c5fd9c01055a91bcb3d50990ac07dbc6a

**Review by conradoplg — APPROVED** [review]
Approving since the fix seem to be correct

**Pull request merged**